### ISSUE TITLE
Implementing a Product Description Shortening and Categorization Pipeline with DSPy

### ISSUE THREAD

**Issue by breakds** [stanfordnlp/dspy#234]
Hello DSPy Team,

I am working on a pipeline in `DSPy` to process Amazon product descriptions by shortening them and classifying them into one of 30 categories. While I understand how to implement the description shortening part, I have some specific questions regarding the categorization aspect:

1. Category Descriptions Inclusion: Should I include short descriptions for each of the 30 categories to assist the model? If so, how should these be incorporated into the module?
2. For 30+ categories, is it necessary to make the compiled module's prompt to include at least one example per category in the training set? What is the best approach to ensure the compiled module does that?
3. Batch Processing for Efficiency: Considering the potential size of the prompt, is there a streamlined way to implement batch input and batch output in `DSPy`? Would composing examples into "batch input / batch output" style be an effective strategy, or is there a better approach within `DSPy` to handle large prompts and reduce overhead?

I appreciate any guidance or insights you can provide for these queries.

Thanks!

Best regards,
-Break

**Comment by okhat** [comment]
Hey @breakds, sorry I missed this!

I'll take a stab at quick high-level answers.

1. One way to do this: consider including descriptions in the docstring of the `dspy.Signature`.
2. No, you can compile with a small number of demonstrations. The LM will typically generalize based on a few (e.g., 8) examples plus the 30 names or descriptions.
3. DSPy's `Evaluate` supports `num_threads`. In general, DSPy is thread-safe, assuming your LM can handle concurrent requests, e.g. `OpenAI` or `TGI` or `VLLM`.

Hope this helps! Let me know if you need further help on this!

In case this doesn't work well, there are more sophisticated ways to do this, but I suspect this is going to be good as-is.

**Comment by breakds** [comment]
Thanks a lot for responding, @okhat!

I think I understand 1 and 2 now. For 3, my concern is that since every time when I ask a question, the whole list of categories (and descriptions) will be in the context, the efficiency (measured by question / (question + context)) is a little too low. By batching I actually mean given the context, I can ask a list of categorization question and expect a list of answers.

**Comment by okhat** [comment]
This is a great point! (It's a great research project to look into *automatic batching* for any task, with LLMs. There was a paper on batching requests a while back, but it was done manually.)

In theory, it should be possible to construct the signature to take `products` (comma separated) and then output `category_per_product` (comma separated). During optimization, your metric kind of needs to be correct / total, instead of just binary.

**Comment by breakds** [comment]
Thanks!

>  During optimization, your metric kind of needs to be correct / total

Like the idea of having such metric as the objective. My understanding is that in this case, all the categories and their description will need to be in the context any way, but the optimizer can help select the best set of examples - is my understanding correct?

**Comment by okhat** [comment]
Yes!
